src/librustc_traits/type_op.rs

```diff
@@ -0,0 +1,117 @@
+        })
+}
+
+fn type_op_normalize<T>(
+    infcx: &InferCtxt<'_, 'gcx, 'tcx>,
+    key: Normalize<'tcx, T>,
+) -> Fallible<InferOk<'tcx, T>>
+where
+    T: fmt::Debug + TypeFoldable<'tcx> + Lift<'gcx>,
+{
+    let Normalize { param_env, value } = key;
+    let Normalized { value, obligations } = infcx
+        .at(&ObligationCause::dummy(), param_env)
+        .normalize(&value)?;
+    Ok(InferOk { value, obligations }) // ugh we should merge these two structs
+}
+
+crate fn type_op_normalize_ty(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    canonicalized: Canonical<'tcx, Normalize<'tcx, Ty<'tcx>>>,
+) -> Result<Lrc<Canonical<'tcx, QueryResult<'tcx, Ty<'tcx>>>>, NoSolution> {
+    tcx.infer_ctxt()
+        .enter_canonical_trait_query(&canonicalized, type_op_normalize)
+}
+
+crate fn type_op_normalize_predicate(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    canonicalized: Canonical<'tcx, Normalize<'tcx, Predicate<'tcx>>>,
+) -> Result<Lrc<Canonical<'tcx, QueryResult<'tcx, Predicate<'tcx>>>>, NoSolution> {
+    tcx.infer_ctxt()
+        .enter_canonical_trait_query(&canonicalized, type_op_normalize)
+}
+
+crate fn type_op_normalize_fn_sig(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    canonicalized: Canonical<'tcx, Normalize<'tcx, FnSig<'tcx>>>,
+) -> Result<Lrc<Canonical<'tcx, QueryResult<'tcx, FnSig<'tcx>>>>, NoSolution> {
+    tcx.infer_ctxt()
+        .enter_canonical_trait_query(&canonicalized, type_op_normalize)
+}
+
+crate fn type_op_normalize_poly_fn_sig(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    canonicalized: Canonical<'tcx, Normalize<'tcx, PolyFnSig<'tcx>>>,
+) -> Result<Lrc<Canonical<'tcx, QueryResult<'tcx, PolyFnSig<'tcx>>>>, NoSolution> {
+    tcx.infer_ctxt()
+        .enter_canonical_trait_query(&canonicalized, type_op_normalize)
+}
+
+crate fn type_op_subtype<'tcx>(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    canonicalized: Canonical<'tcx, Subtype<'tcx>>,
+) -> Result<Lrc<Canonical<'tcx, QueryResult<'tcx, ()>>>, NoSolution> {
+    tcx.infer_ctxt().enter_canonical_trait_query(
+        &canonicalized,
+        |infcx,
+         Subtype {
+             param_env,
+             sub,
+             sup,
+         }| {
+            Ok(infcx
+                .at(&ObligationCause::dummy(), param_env)
+                .sup(sup, sub)?)
+        },
+    )
+}
+
+crate fn type_op_prove_predicate<'tcx>(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    canonicalized: Canonical<'tcx, ProvePredicate<'tcx>>,
+) -> Result<Lrc<Canonical<'tcx, QueryResult<'tcx, ()>>>, NoSolution> {
+    tcx.infer_ctxt()
+        .enter_canonical_trait_query(&canonicalized, |_infcx, key| {
+            let ProvePredicate {
+                param_env,
+                predicate,
+            } = key;
+            Ok(InferOk {
+                value: (),
+                obligations: vec![Obligation::new(
+                    ObligationCause::dummy(),
+                    param_env,
+                    predicate,
+                )],
```

Comment:
Oookay turns out GitHub deleted replies to comments... I was only wanting to move it.

@nikomatsakis replied to the original version of this (found in my email):
> Hmm. That's not a bad idea. I've been trying to encapsulate the FulfillmentContext to some degree, since it's part of the "old trait system" and not really a part of the new one, and for most of the methods they have an InferOk all setup to return. But the inefficiency here does annoy me so....